demos/src/mqtt_demo.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+
+#include "mqtt.h"
+#include "transport_plaintext.h"
+
+#define SERVER    "test.mosquitto.org"
+#define PORT      1883
+
+#define NETWORK_BUFFER_SIZE    ( 1024U )
+
+#define CLIENT_IDENTIFIER           "testclient"
+#define CLIENT_IDENTIFIER_LENGTH    ( ( uint16_t ) ( sizeof( CLIENT_IDENTIFIER ) - 1 ) )
+
+static void eventCallback( MQTTContext_t * pContext, MQTTPacketInfo_t * pPacketInfo )
+{
+
+}
+
+static uint32_t getTime( void )
+{
+    return 0;
+}
+
+static int establishMqttSession( MQTTContext_t * pContext, int tcpSocket )
+{
+    int status = EXIT_SUCCESS;
+    MQTTStatus_t mqttStatus;
+    MQTTConnectInfo_t connectInfo;
+
+    /* These members are not copied into the context. They must remain in scope
+     * for the lifetime of the context. */
+    static MQTTTransportInterface_t transport;
```

Comment:
Why is that - these are just function pointers and should be copied into the context. The only thing that should not be copied is the buffer: 

```c
MQTTTransportInterface_t transport;
MQTTFixedBuffer_t networkBuffer;
MQTTApplicationCallbacks_t callbacks;
static uint8_t buffer[ NETWORK_BUFFER_SIZE ];
```

Reply:
The structs are on the call stack of establishMqttConnection.  MQTT_Init remembers the pointer to them.  Those structs must remain in scope because the MQTT context will use them after establishMqttConnection returns.